cloudflare/resource_cloudflare_waf_override.go

```diff
@@ -0,0 +1,221 @@
+
+	override, err := client.WAFOverride(zoneID, d.Id())
+	if err != nil {
+		if strings.Contains(err.Error(), "wafuriconfig.api.not_found") {
+			log.Printf("[INFO] WAF override %s no longer exists", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("failed to find WAF override %s: %s", d.Id(), err)
+	}
+
+	d.Set("zone_id", zoneID)
+	d.Set("urls", override.URLs)
+	d.Set("rules", override.Rules)
+	d.Set("paused", override.Paused)
+	d.Set("description", override.Description)
+	d.Set("priority", override.Priority)
+
+	if len(override.Groups) != 0 {
+		d.Set("groups", override.Groups)
+	}
+
+	if len(override.RewriteAction) != 0 {
+		d.Set("rewrite_action", override.RewriteAction)
+	}
+
+	d.Set("override_id", override.ID)
+
+	return nil
+}
+
+func resourceCloudflareWAFOverrideCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	zoneID := d.Get("zone_id").(string)
+	newOverride, _ := buildWAFOverride(d)
+
+	override, err := client.CreateWAFOverride(zoneID, newOverride)
+	if err != nil {
+		return fmt.Errorf("failed to create WAF override: %s", err)
+	}
+
+	d.SetId(override.ID)
+
+	return resourceCloudflareWAFOverrideRead(d, meta)
+}
+
+func resourceCloudflareWAFOverrideUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	zoneID := d.Get("zone_id").(string)
+	overrideID := d.Get("override_id").(string)
+	updatedOverride, _ := buildWAFOverride(d)
+
+	_, err := client.UpdateWAFOverride(zoneID, overrideID, updatedOverride)
+	if err != nil {
+		return fmt.Errorf("failed to update WAF override: %s", err)
+	}
+
+	return resourceCloudflareWAFOverrideRead(d, meta)
+}
+
+func resourceCloudflareWAFOverrideDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	overrideID := d.Get("override_id").(string)
+	zoneID := d.Get("zone_id").(string)
+
+	err := client.DeleteWAFOverride(zoneID, overrideID)
+	if err != nil {
+		return fmt.Errorf("failed to delete WAF override ID %s: %s", overrideID, err)
+	}
+
+	return resourceCloudflareWAFOverrideRead(d, meta)
+}
+
+func resourceCloudflareWAFOverrideImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	idAttr := strings.SplitN(d.Id(), "/", 2)
+
+	if len(idAttr) != 2 {
+		return nil, fmt.Errorf("invalid id (\"%s\") specified, should be in format \"zoneID/WAFOverrideID\"", d.Id())
+	}
+
+	zoneID, WAFOverrideID := idAttr[0], idAttr[1]
+
+	log.Printf("[DEBUG] Importing WAF override: id %s for zone %s", WAFOverrideID, zoneID)
+
+	d.Set("zone_id", zoneID)
+	d.Set("override_id", WAFOverrideID)
+	d.SetId(WAFOverrideID)
+
+	resourceCloudflareWAFOverrideRead(d, meta)
+
+	return []*schema.ResourceData{d}, nil
+}
+
+// buildWAFOverride centralises the creation of a WAFOverride struct which can
+// be reused between Create and Update methods to ensure consistent building of
+// the values.
+func buildWAFOverride(d *schema.ResourceData) (cloudflare.WAFOverride, error) {
```

Comment:
We don't yet emit `error` anywhere in here but I'd rather return it now and have it already accounted for then try to introduce a breaking change later on.